pbsmetrics/go_metrics.go

```diff
@@ -121,6 +151,186 @@ func registerAdapterMetrics(registry metrics.Registry, adapterOrAccount string,
+	var ok bool
+
+	me.accountMetricsRWMutex.RLock()
+	am, ok = me.accountMetrics[id]
+	me.accountMetricsRWMutex.RUnlock()
+
+	if ok {
+		return am
+	}
+
+	me.accountMetricsRWMutex.Lock()
+	// Made sure to use defer as we have two exit points: we want to unlock the mutex as quickly as possible.
+	defer me.accountMetricsRWMutex.Unlock()
+
+	am, ok = me.accountMetrics[id]
+	if ok {
+		// Unlock and return as quickly as possible
+		return am
+	}
+	am = &accountMetrics{}
+	am.requestMeter = metrics.GetOrRegisterMeter(fmt.Sprintf("account.%s.requests", id), me.metricsRegistry)
+	am.bidsReceivedMeter = metrics.GetOrRegisterMeter(fmt.Sprintf("account.%s.bids_received", id), me.metricsRegistry)
+	am.priceHistogram = metrics.GetOrRegisterHistogram(fmt.Sprintf("account.%s.prices", id), me.metricsRegistry, metrics.NewExpDecaySample(1028, 0.015))
+	am.adapterMetrics = make(map[openrtb_ext.BidderName]*AdapterMetrics, len(me.exchanges))
+	for _, a := range me.exchanges {
+		am.adapterMetrics[a] = makeBlankAdapterMetrics(me.metricsRegistry, a)
+		registerAdapterMetrics(me.metricsRegistry, fmt.Sprintf("account.%s", id), string(a), am.adapterMetrics[a])
+	}
+
+	me.accountMetrics[id] = am
+
+	return am
+}
+
+// Implement the MetricsEngine interface
+
+// RecordRequest implements a part of the MetricsEngine interface
+func (me *Metrics) RecordRequest(labels Labels) {
+	me.RequestMeter.Mark(1)
+	if labels.Source == DemandApp {
+		me.AppRequestMeter.Mark(1)
+	} else {
+		if labels.Browser == BrowserSafari {
+			me.SafariRequestMeter.Mark(1)
+			if labels.CookieFlag == CookieFlagNo {
+				me.SafariNoCookieMeter.Mark(1)
+			}
+		}
+		if labels.CookieFlag == CookieFlagNo {
+			// NOTE: Old behavior was log me.AMPNoCookieMeter here for AMP requests.
+			// AMP is still new and OpenRTB does not do this, so changing to match
+			// OpenRTB endpoint
+			me.NoCookieMeter.Mark(1)
+		}
+	}
+	switch labels.RType {
+	case ReqTypeORTB2:
+		me.ORTBRequestMeter.Mark(1)
+	case ReqTypeAMP:
+		me.AmpRequestMeter.Mark(1)
+	}
+	if labels.RequestStatus == RequestStatusErr {
+		me.ErrorMeter.Mark(1)
+	}
+	// Handle the account metrics now.
+	am := me.getAccountMetrics(labels.PubID)
+	am.requestMeter.Mark(1)
+}
+
+// RecordRequestTime implements a part of the MetricsEngine interface. The calling code is responsible
+// for determining the call duration.
+func (me *Metrics) RecordRequestTime(labels Labels, length time.Duration) {
+	// Only record times for successful requests, as we don't have labels to screen out bad requests.
+	if labels.RequestStatus == RequestStatusOK {
+		me.RequestTimer.Update(length)
+	}
+}
+
+// RecordAdapterRequest implements a part of the MetricsEngine interface
+func (me *Metrics) RecordAdapterRequest(labels Labels) {
+	_, ok := me.AdapterMetrics[labels.Adapter]
+	if !ok {
+		glog.Errorf("Trying to run adapter metrics on %s: adapter metrics not found", string(labels.Adapter))
+		return
+	}
+	// Adapter metrics
+	am := me.AdapterMetrics[labels.Adapter]
```

Comment:
Tiny simplification: you could move this to `am, ok := me.AdapterMetrics[labels.Adapter]` and save a map lookup.

Same goes for some other functions in this file.

Reply:
Good catch, thanks.